scripts/preprocess/process_zip_to_shp.R

```diff
@@ -25,18 +25,113 @@ read_shp_zip <- function(zipfile) {
 }
 
 #' @param sp the full spatial object to be altered, w/ STATEFP attribute
-#' @param fips a character vector of fips to be scaled
+#' @param ... named character argument for fields in `sp` to be scaled
 #' @param scale a scale factor to apply to fips
 #' @return an `sp` similar to the input, but with the specified fips scaled according to `scale` parameter
-scale_shifted_shps <- function(sp, fips, scale){
-  
-  toshift_sp <- sp[sp@data$STATEFP %in% fips, ]
-  toshift_cent <- rgeos::gCentroid(toshift_sp, byid=F)@coords
-  toshift_scale <- max(apply(bbox(toshift_sp), 1, diff)) * scale
-  obj <- maptools::elide(toshift_sp, scale=toshift_scale, center=toshift_cent, bb = bbox(toshift_sp))
-  new_cent <- rgeos::gCentroid(obj, byid=FALSE)@coords
-  obj <- maptools::elide(obj, shift=c(toshift_cent-new_cent))
-  proj4string(obj) <- proj4string(sp)
-  sp_out <- rbind(sp[!(sp@data$STATEFP %in% fips), ], obj)
+scale_shifted_shps <- function(sp, ..., scale){
+  
+  args <- list(...)
+  field <- names(args)
+  if (length(field) != 1){
+    stop(args, ' was not valid')
+  }
+  values <- args[[1]]
+  if (length(values) != length(scale)){
+    stop('number of values in ',field,' must equal the length of scale')
+  }
+  for (i in 1:length(values)){
+    toshift_sp <- sp[sp@data[[field]] == values[i], ]
+    toshift_cent <- rgeos::gCentroid(toshift_sp, byid=F)@coords
+    toshift_scale <- max(apply(bbox(toshift_sp), 1, diff)) * scale[i]
+    obj <- maptools::elide(toshift_sp, scale=toshift_scale, center=toshift_cent, bb = bbox(toshift_sp))
+    new_cent <- rgeos::gCentroid(obj, byid=FALSE)@coords
+    obj <- maptools::elide(obj, shift=c(toshift_cent-new_cent))
+    proj4string(obj) <- proj4string(sp)
+    sp_out <- rbind(sp[!(sp@data[[field]] == values[i]), ], obj)
+    sp <- sp_out
+  }
+  
   return(sp_out)
+}
+
+geomcoll_to_poly <- function(sfg){
+  classes <- sapply(seq_len(length(sfg)), function(x) class(sfg[x][[1]])[2])
+  is_poly <- grepl("*POLYGON", classes)
+  if (any(is_poly)){
+    sfg[[which(is_poly)[1L]]]
+  } else {
+    NULL
+  }
+}
+
+linepoint_to_poly <- function(sfg){
+  # assumes lat/lon units
+  dtol <- 0.01
+  
+  # coerce into polygon
+  sfg_poly <- st_buffer(sfg, dtol)
+  
+  #simplify "rounded" polygon
+  st_simplify(sfg_poly, dTolerance = dtol/2)
+}
+
+
+rescale_write_shps <- function(filename_out, topojson_filename, ..., scale, quantize, coerce_dropped = FALSE){
+  shps <- topojson_read(topojson_filename) %>% 
+    st_as_sf() %>% 
+    st_make_valid() 
+  
+  sfc_objects <- sf::st_geometry(shps)
+  
+  types <- vapply(sfc_objects, function(x) {
+    class(x)[2]
+  }, "")
+  
+  drop_counties <- grepl("*LINE", types) | grepl("*POINT", types)
+  geo_col <- grepl("*GEOMETRYCOLLECTION", types) # test that there aren't others?
+  
+  for (j in which(geo_col)){
+    sfg <- geomcoll_to_poly(sfg = sfc_objects[[j]])
```

Comment:
simple feature geometry. The reason they are the same (but yes, I see why this is confusing) is because it is returning the same thing - just coercing GEOMETRYCOLLECTION into MULTIPOLYGON (geocollections have multiple different feature types, including polys, lines, and points)